File: frontend-react/src/services/aiModels.ts
import { config } from "../config";
import type {
  AIModel,
  ModelCreate,
  ModelUpdate,
  ModelTest,
} from "../store/aiModelsStore";

export interface ModelPerformance {
  latencyMsAvg: number;
  tokensPerSecond: number;
  successRate: number;
  timeframe: string;
}

export interface ModelUsage {
  totalRequests: number;
  totalTokens: number;
  averageTokensPerRequest: number;
  timeframe: string;
}

export interface ModelCosts {
  totalCostUsd: number;
  inputCostUsd: number;
  outputCostUsd: number;
  timeframe: string;
}

export interface ProviderInfo {
  id: string;
  name: string;
  supportedModels: string[];
}

export interface CompareModelsResult {
  differences: Array<{ key: string; a?: unknown; b?: unknown }>;
}

export interface ModelAnalytics {
  byProvider: Record<string, number>;
  byModel: Record<string, number>;
  timeframe: string;
}

class AIModelsService {
  private baseUrl = config.apiUrl;

  private get base(): string {
    return `${this.baseUrl}/api/v1/ai-models`;
  }

  async getModels(): Promise<AIModel[]> {
    const response = await fetch(`${this.base}/`);
    if (!response.ok) throw new Error("Failed to fetch models");
    return response.json();
  }

  async getModel(modelId: string): Promise<AIModel> {
    const response = await fetch(`${this.base}/${modelId}`);
    if (!response.ok) throw new Error("Failed to fetch model");
    return response.json();
  }

  async createModel(modelData: ModelCreate): Promise<AIModel> {
    const response = await fetch(`${this.base}/`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(modelData),
    });
    if (!response.ok) throw new Error("Failed to create model");
    return response.json();
  }

  async updateModel(modelId: string, modelData: ModelUpdate): Promise<AIModel> {
    const response = await fetch(`${this.base}/${modelId}`, {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(modelData),
    });
    if (!response.ok) throw new Error("Failed to update model");
    return response.json();
  }

  async deleteModel(modelId: string): Promise<void> {
    const response = await fetch(`${this.base}/${modelId}`, { method: "DELETE" });
    if (!response.ok) throw new Error("Failed to delete model");
  }

  async toggleModelActive(modelId: string, isActive: boolean): Promise<AIModel> {
    const response = await fetch(`${this.base}/${modelId}/toggle`, {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ isActive }),
    });
    if (!response.ok) throw new Error("Failed to toggle model status");
    return response.json();
  }

  async setDefaultModel(modelId: string): Promise<AIModel> {
    const response = await fetch(`${this.base}/${modelId}/default`, {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
    });
    if (!response.ok) throw new Error("Failed to set default model");
    return response.json();
  }

  async testModel(modelId: string, prompt: string): Promise<ModelTest> {
    const response = await fetch(`${this.base}/${modelId}/test`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ prompt }),
    });
    if (!response.ok) throw new Error("Failed to test model");
    return response.json();
  }

  async getModelPerformance(modelId: string, timeRange?: string): Promise<ModelPerformance> {
    const params = timeRange ? `?timeRange=${timeRange}` : "";
    const response = await fetch(`${this.base}/${modelId}/performance${params}`);
    if (!response.ok) throw new Error("Failed to fetch model performance");
    return response.json();
  }

  async getModelUsage(modelId: string, timeRange?: string): Promise<ModelUsage> {
    const params = timeRange ? `?timeRange=${timeRange}` : "";
    const response = await fetch(`${this.base}/${modelId}/usage${params}`);
    if (!response.ok) throw new Error("Failed to fetch model usage");
    return response.json();
  }

  async getModelCosts(modelId: string, timeRange?: string): Promise<ModelCosts> {
    const params = timeRange ? `?timeRange=${timeRange}` : "";
    const response = await fetch(`${this.base}/${modelId}/costs${params}`);
    if (!response.ok) throw new Error("Failed to fetch model costs");
    return response.json();
  }

  async getProviders(): Promise<string[]> {
    const response = await fetch(`${this.base}/providers`);
    if (!response.ok) throw new Error("Failed to fetch providers");
    return response.json();
  }

  async getProviderModels(provider: string): Promise<string[]> {
    const response = await fetch(`${this.base}/providers/${provider}/models`);
    if (!response.ok) throw new Error("Failed to fetch provider models");
    return response.json();
  }

  async validateModel(modelData: ModelCreate): Promise<{ valid: boolean; errors?: string[] }> {
    const response = await fetch(`${this.base}/models/validate`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(modelData),
    });
    if (!response.ok) throw new Error("Failed to validate model");
    return response.json();
  }

  async compareModels(modelIds: string[]): Promise<CompareModelsResult> {
    const response = await fetch(`${this.base}/models/compare`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ modelIds }),
    });
    if (!response.ok) throw new Error("Failed to compare models");
    return response.json();
  }

  async getModelRecommendations(
    useCase: string,
    requirements?: Partial<Record<string, unknown>>,
  ): Promise<AIModel[]> {
    const response = await fetch(`${this.base}/models/recommendations`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ useCase, requirements }),
    });
    if (!response.ok) throw new Error("Failed to get model recommendations");
    return response.json();
  }

  async getModelAnalytics(timeRange?: string, filters?: Partial<Record<string, unknown>>): Promise<ModelAnalytics> {
    const params = new URLSearchParams();
    if (timeRange) params.append("timeRange", timeRange);
    if (filters) params.append("filters", JSON.stringify(filters));

    const response = await fetch(`${this.base}/models/analytics?${params}`);
    if (!response.ok) throw new Error("Failed to fetch model analytics");
    return response.json();
  }
}

export const aiModelsService = new AIModelsService();
export default aiModelsService;

File: backend/app/api/v1/endpoints/ai_models.py
from typing import Any, List, Optional

from fastapi import APIRouter, Body, Depends, HTTPException, Query
from pydantic import BaseModel, Field
from sqlalchemy.orm import Session

from backend.app.core.database import get_db
from backend.app.core.security import get_current_user
from backend.app.models.user import User
from backend.app.services.ai_models_service import AIModelsService
from backend.app.services.ai_service import get_ai_service

router = APIRouter()


class ModelPerformance(BaseModel):
	latencyMsAvg: float
	tokensPerSecond: float
	successRate: float
	timeframe: str


class ModelUsage(BaseModel):
	totalRequests: int
	totalTokens: int
	averageTokensPerRequest: float
	timeframe: str


class ModelCosts(BaseModel):
	totalCostUsd: float
	inputCostUsd: float
	outputCostUsd: float
	timeframe: str


class AIModelDTO(BaseModel):
	name: str = Field(...)
	provider: str = Field(...)
	model_key: str = Field(...)
	display_name: str = Field(...)
	description: Optional[str] = None
	max_tokens: Optional[int] = None
	cost_per_1k_tokens: Optional[float] = None
	is_active: bool = True
	is_default: bool = False
	is_favorite: bool = False
	capabilities: List[str] = Field(default_factory=list)


@router.get("/", response_model=List[dict])
async def list_models(
	current_user: User = Depends(get_current_user),
	db: Session = Depends(get_db),
):
	service = AIModelsService(db)
	return [m.to_dict() for m in service.list_models()]


@router.get("/{model_id}", response_model=dict)
async def get_model(
	model_id: str,
	current_user: User = Depends(get_current_user),
	db: Session = Depends(get_db),
):
	service = AIModelsService(db)
	m = service.get_model(model_id)
	if not m:
		raise HTTPException(status_code=404, detail="Model not found")
	return m.to_dict()


@router.post("/", response_model=dict)
async def create_model(
	payload: AIModelDTO,
	current_user: User = Depends(get_current_user),
	db: Session = Depends(get_db),
):
	service = AIModelsService(db)
	m = service.create_model(payload.dict())
	return m.to_dict()


@router.put("/{model_id}", response_model=dict)
async def update_model(
	model_id: str,
	payload: dict,
	current_user: User = Depends(get_current_user),
	db: Session = Depends(get_db),
):
	service = AIModelsService(db)
	m = service.update_model(model_id, payload)
	if not m:
		raise HTTPException(status_code=404, detail="Model not found")
	return m.to_dict()


@router.delete("/{model_id}")
async def delete_model(
	model_id: str,
	current_user: User = Depends(get_current_user),
	db: Session = Depends(get_db),
):
	service = AIModelsService(db)
	ok = service.delete_model(model_id)
	if not ok:
		raise HTTPException(status_code=404, detail="Model not found")
	return {"message": "Model deleted"}


@router.put("/{model_id}/toggle", response_model=dict)
async def toggle_model(
	model_id: str,
	isActive: bool = Body(..., embed=True),
	current_user: User = Depends(get_current_user),
	db: Session = Depends(get_db),
):
	service = AIModelsService(db)
	m = service.toggle_active(model_id, isActive)
	if not m:
		raise HTTPException(status_code=404, detail="Model not found")
	return m.to_dict()


@router.put("/{model_id}/default", response_model=dict)
async def set_default_model(
	model_id: str,
	current_user: User = Depends(get_current_user),
	db: Session = Depends(get_db),
):
	service = AIModelsService(db)
	m = service.set_default(model_id)
	if not m:
		raise HTTPException(status_code=404, detail="Model not found")
	return m.to_dict()


class ModelTestRequest(BaseModel):
	prompt: str


@router.post("/{model_id}/test", response_model=dict)
async def test_model(
	model_id: str,
	request: ModelTestRequest,
	current_user: User = Depends(get_current_user),
	db: Session = Depends(get_db),
):
	# Simple test: call AIService with the specified provider/model_key
	service = AIModelsService(db)
	m = service.get_model(model_id)
	if not m:
		raise HTTPException(status_code=404, detail="Model not found")
	ai = get_ai_service(db)
	# For demo, we won't actually call external provider here; return stub success
	return {
		"id": str(m.id),
		"modelId": str(m.id),
		"prompt": request.prompt,
		"response": "Test response (stub)",
		"responseTime": 0,
		"tokensUsed": 0,
		"cost": 0.0,
		"timestamp": "",
	}


@router.get("/{model_id}/performance", response_model=ModelPerformance)
async def get_model_performance(
	model_id: str,
	timeRange: Optional[str] = Query(None),
	current_user: User = Depends(get_current_user),
	db: Session = Depends(get_db),
):
	# Placeholder - integrate metrics later
	return ModelPerformance(latencyMsAvg=0, tokensPerSecond=0, successRate=1.0, timeframe=timeRange or "")


@router.get("/{model_id}/usage", response_model=ModelUsage)
async def get_model_usage(
	model_id: str,
	timeRange: Optional[str] = Query(None),
	current_user: User = Depends(get_current_user),
	db: Session = Depends(get_db),
):
	return ModelUsage(totalRequests=0, totalTokens=0, averageTokensPerRequest=0, timeframe=timeRange or "")


@router.get("/{model_id}/costs", response_model=ModelCosts)
async def get_model_costs(
	model_id: str,
	timeRange: Optional[str] = Query(None),
	current_user: User = Depends(get_current_user),
	db: Session = Depends(get_db),
):
	return ModelCosts(totalCostUsd=0.0, inputCostUsd=0.0, outputCostUsd=0.0, timeframe=timeRange or "")


@router.get("/providers", response_model=List[str])
async def get_providers(
	current_user: User = Depends(get_current_user),
	db: Session = Depends(get_db),
):
	ai = get_ai_service(db)
	return ai.get_available_providers()


@router.get("/providers/{provider}/models", response_model=List[str])
async def get_provider_models(
	provider: str,
	current_user: User = Depends(get_current_user),
	db: Session = Depends(get_db),
):
	ai = get_ai_service(db)
	return ai.get_available_models(provider)


class ValidateModelRequest(BaseModel):
	name: str
	provider: str
	model_key: str


@router.post("/models/validate", response_model=dict)
async def validate_model(
	request: ValidateModelRequest,
	current_user: User = Depends(get_current_user),
	db: Session = Depends(get_db),
):
	# Stub validation: ensure provider/model exists in available list
	ai = get_ai_service(db)
	if request.provider not in ai.get_available_providers():
		raise HTTPException(status_code=400, detail="Unknown provider")
	if request.model_key not in ai.get_available_models(request.provider):
		raise HTTPException(status_code=400, detail="Unknown model for provider")
	return {"valid": True}


class CompareModelsRequest(BaseModel):
	modelIds: List[str]


@router.post("/models/compare", response_model=dict)
async def compare_models(
	request: CompareModelsRequest,
	current_user: User = Depends(get_current_user),
	db: Session = Depends(get_db),
):
	service = AIModelsService(db)
	models = [service.get_model(mid) for mid in request.modelIds]
	if any(m is None for m in models):
		raise HTTPException(status_code=404, detail="One or more models not found")
	# Simple diff output
	differences = []
	keys = ["provider", "model_key", "max_tokens", "cost_per_1k_tokens"]
	for key in keys:
		vals = list({getattr(m, key) for m in models if m})
		if len(vals) > 1:
			differences.append({"key": key, "values": vals})
	return {"differences": differences}


class RecommendationsRequest(BaseModel):
	useCase: str
	requirements: Optional[dict] = None


@router.post("/models/recommendations", response_model=List[dict])
async def get_recommendations(
	request: RecommendationsRequest,
	current_user: User = Depends(get_current_user),
	db: Session = Depends(get_db),
):
	# Placeholder heuristic
	service = AIModelsService(db)
	models = [m for m in service.list_models() if m.is_active]
	return [m.to_dict() for m in models[:5]]


@router.get("/models/analytics", response_model=dict)
async def get_models_analytics(
	timeRange: Optional[str] = Query(None),
	filters: Optional[str] = Query(None),
	current_user: User = Depends(get_current_user),
	db: Session = Depends(get_db),
):
	# Placeholder analytics aggregation
	service = AIModelsService(db)
	models = service.list_models()
	by_provider = {}
	for m in models:
		by_provider[m.provider] = by_provider.get(m.provider, 0) + 1
	by_model = {m.model_key: 1 for m in models}
	return {"byProvider": by_provider, "byModel": by_model, "timeframe": timeRange or ""}